File: extra/executor.ts
import { compileRitual } from './compiler';

async function executeRitual(filePath: string, context: any) {
  const { ir, graph } = compileRitual(filePath);
  const state = { ...context, ritualState: ir.awareness.state };
  const auditLog = [];

  for (const node of graph.nodes) {
    switch (node.type) {
      case 'start':
        auditLog.push({ action: 'start', agent: node.data.agent });
        break;
      case 'check':
        if (!state[node.data.condition]) {
          throw new Error(`Precondition failed: ${node.data.condition}`);
        }
        auditLog.push({ action: 'check', condition: node.data.condition });
        break;
      case 'static':
      case 'adaptive':
      case 'hormonal':
        if (!evaluateDefense(node.data.logic, state, node.data.params)) {
          throw new Error(`Defense failed: ${node.data.logic}`);
        }
        auditLog.push({ action: 'defense', type: node.type, logic: node.data.logic });
        break;
      case 'action':
        state[node.id] = node.data.action;
        auditLog.push({ action: 'response', response: node.data.action });
        break;
      case 'loop':
        let depth = 0;
        while (depth < node.data.maxDepth && evaluateCondition(node.data.condition, state)) {
          await executeRitual(filePath, state);
          depth++;
        }
        auditLog.push({ action: 'recursive', depth });
        break;
      case 'log':
        auditLog.push({ action: 'audit', ...node.data });
        break;
    }
  }

  return { state, auditLog };
}

// Mock evaluation functions (replace with real logic)
function evaluateDefense(logic: string, state: any, params: any): boolean {
  return true; // Simulate defense check
}

function evaluateCondition(condition: string, state: any): boolean {
  return true; // Simulate condition check
}

export { executeRitual };

File: extra/executor.rs
use serde_json::{Value};
use std::collections::HashMap;

#[derive(Debug)]
struct AuditLog {
    action: String,
    data: String,
    timestamp: String,
}

pub struct RitualExecutor {
    state: HashMap<String, HashMap<String, bool>>,
    logs: HashMap<String, Vec<AuditLog>>,
}

impl RitualExecutor {
    pub fn new() -> Self {
        RitualExecutor {
            state: HashMap::new(),
            logs: HashMap::new(),
        }
    }

    pub fn execute_ritual(&mut self, ritual: &Value) -> Result<(), String> {
        let ritual_id = ritual["ritual"].as_str().ok_or("Invalid ritual ID")?;
        let preconditions = ritual["requires"].as_array().ok_or("Invalid preconditions")?;
        let defenses = ritual["defense"].as_array().ok_or("Invalid defenses")?;
        let responses = ritual["response"].as_array().ok_or("Invalid responses")?;
        let recursive = ritual["recursive"].as_object().ok_or("Invalid recursive")?;
        let max_depth = recursive["max_depth"].as_u64().unwrap_or(10) as usize;

        for p in preconditions {
            let key = p.as_str().ok_or("Invalid precondition")?;
            if !self.state.get(ritual_id).and_then(|s| s.get(key)).unwrap_or(&false) {
                return Err(format!("Precondition failed: {}", key));
            }
            self.logs.entry(ritual_id.to_string()).or_insert(vec![]).push(AuditLog {
                action: "check".to_string(),
                data: key.to_string(),
                timestamp: "2025-05-22T10:15:00Z".to_string(),
            });
        }

        for d in defenses {
            // Mock defense check
            self.logs.entry(ritual_id.to_string()).or_insert(vec![]).push(AuditLog {
                action: "defense".to_string(),
                data: d["logic"].as_str().unwrap_or("").to_string(),
                timestamp: "2025-05-22T10:15:00Z".to_string(),
            });
        }

        for r in responses {
            let action = r.as_str().ok_or("Invalid response")?;
            self.state.entry(ritual_id.to_string()).or_insert(HashMap::new()).insert(action.to_string(), true);
            self.logs.entry(ritual_id.to_string()).or_insert(vec![]).push(AuditLog {
                action: "response".to_string(),
                data: action.to_string(),
                timestamp: "2025-05-22T10:15:00Z".to_string(),
            });
        }

        if recursive["enabled"].as_bool().unwrap_or(false) && max_depth > 0 {
            // Mock condition check
            if true {
                self.execute_ritual(ritual)?;
            }
        }

        self.logs.entry(ritual_id.to_string()).or_insert(vec![]).push(AuditLog {
            action: "audit".to_string(),
            data: "completed".to_string(),
            timestamp: "2025-05-22T10:15:00Z".to_string(),
        });

        Ok(())
    }
}